graph/mongo/indexed_linksto.go

```diff
@@ -0,0 +1,278 @@
+	constraint := bson.M{}
+	for _, set := range it.lset {
+		var s []string
+		for _, v := range set.Values {
+			s = append(s, it.qs.NameOf(v))
+		}
+		constraint[set.Dir.String()] = bson.M{"$in": s}
+		if len(s) == 1 {
+			constraint[set.Dir.String()] = s[0]
+		}
+	}
+	return constraint
+}
+
+func (it *LinksTo) buildIteratorFor(d quad.Direction, val graph.Value) *mgo.Iter {
+	name := it.qs.NameOf(val)
+	constraint := it.buildConstraint()
+	constraint[d.String()] = name
+	return it.qs.db.C(it.collection).Find(constraint).Iter()
+}
+
+func (it *LinksTo) UID() uint64 {
+	return it.uid
+}
+
+func (it *LinksTo) Tagger() *graph.Tagger {
+	return &it.tags
+}
+
+// Return the direction under consideration.
+func (it *LinksTo) Direction() quad.Direction { return it.dir }
+
+// Tag these results, and our subiterator's results.
+func (it *LinksTo) TagResults(dst map[string]graph.Value) {
+	for _, tag := range it.tags.Tags() {
+		dst[tag] = it.Result()
+	}
+
+	for tag, value := range it.tags.Fixed() {
+		dst[tag] = value
+	}
+
+	it.primaryIt.TagResults(dst)
+}
+
+// DEPRECATED
+func (it *LinksTo) ResultTree() *graph.ResultTree {
+	tree := graph.NewResultTree(it.Result())
+	tree.AddSubtree(it.primaryIt.ResultTree())
+	return tree
+}
+
+// Optimize the LinksTo, by replacing it if it can be.
+func (it *LinksTo) Optimize() (graph.Iterator, bool) {
+	return it, false
+}
+
+func (it *LinksTo) Next() bool {
+	var result struct {
+		ID      string  `bson:"_id"`
+		Added   []int64 `bson:"Added"`
+		Deleted []int64 `bson:"Deleted"`
+	}
+	graph.NextLogIn(it)
+	it.runstats.Next += 1
+	if it.nextIt != nil && it.nextIt.Next(&result) {
+		it.runstats.ContainsNext += 1
+		if it.collection == "quads" && len(result.Added) <= len(result.Deleted) {
+			return it.Next()
+		}
+		it.result = result.ID
+		return graph.NextLogOut(it, it.result, true)
+	}
+
+	if it.nextIt != nil {
+		// If there's an error in the 'next' iterator, we save it and we're done.
+		it.err = it.nextIt.Err()
+		if it.err != nil {
+			return false
+		}
+
+	}
+	// Subiterator is empty, get another one
+	if !graph.Next(it.primaryIt) {
+		// Possibly save error
+		it.err = it.primaryIt.Err()
+
+		// We're out of nodes in our subiterator, so we're done as well.
+		return graph.NextLogOut(it, 0, false)
+	}
+	if it.nextIt != nil {
+		it.nextIt.Close()
+	}
+	it.nextIt = it.buildIteratorFor(it.dir, it.primaryIt.Result())
+
+	// Recurse -- return the first in the next set.
+	return it.Next()
+}
+
+func (it *LinksTo) Err() error {
+	return it.err
+}
+
+func (it *LinksTo) Result() graph.Value {
+	return it.result
+}
+
+func (it *LinksTo) Close() error {
+	var err error
+	if it.nextIt != nil {
+		err = it.nextIt.Close()
+	}
+
+	_err := it.primaryIt.Close()
+	if _err != nil && err == nil {
+		err = _err
+	}
+
+	return err
+}
+
+func (it *LinksTo) NextPath() bool {
+	ok := it.primaryIt.NextPath()
+	if !ok {
+		it.err = it.primaryIt.Err()
+	}
+	return ok
+}
+
+var mongoIndexedLinksToType graph.Type
```

Comment:
s/mongoIndexedLinksToType/linksToType/g

This will read as mongo.linksToType. The indexation is an implementation detail.